### PR TITLE
APP-2019 Export cancellation from admin ui

### PR DESCRIPTION
**Pull request by pec1985**

### CODE REVIEW

**Comment by xzandrew** [comment]
We would need to adjust the checkpoint folders a bit to make sure they are kept in sync if the export is cancelled in the middle.

File: pkg/forcekill/forcekill.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+	// this, _ := pps.FindProcess(pid)
+	// var names []string
+	// if this != nil {
+	// 	names = []string{this.Executable()}
+	// }
+
+	for {
+		var found bool
+		for _, p := range prs {
+			if p.PPid() == pid {
```
Comment (xzandrew) [inline pkg/forcekill/forcekill.go]: If I'm reading this right, it would only kill 1 process per parent, so if child has multiple processes started only 1 will be killed.

File: pkg/fsconf/fsconf.go [hunk 2]
```diff
@@ -16,19 +16,21 @@ type Locs struct {
...
 
 	ServiceRunCrashes string
 
 	// Special files
 	Config2 string // new config that is populated from enroll, not for manual editing
 
 	// LastProcessedFile stores timestamps or other data to mark last processed objects
-	LastProcessedFile string
+	LastProcessedFile    string
+	LastProcessedFileTmp string
```
Comment (xzandrew) [inline pkg/fsconf/fsconf.go]: Rename LastProcessedFileTmp to LastProcessedFileBackup

File: cmd/cmdexport/export.go [hunk 3]
```diff
@@ -87,11 +90,36 @@ func newExport(opts Opts) (*export, error) {
...
+	if fileutil.FileExists(s.Locs.State) {
+
+		// if there is a tmp processed file and checkpoint dir, it means that the
+		// export was killed. If so, delete the current processed file and checkpoint
+		// dir and copy the those temps.
+		// if there are no temps, create them and defer delete
 
+		if fileutil.FileExists(s.Locs.LastProcessedFileTmp) {
+			os.RemoveAll(s.Locs.LastProcessedFile)
+			copy.Copy(s.Locs.LastProcessedFileTmp, s.Locs.LastProcessedFile)
```
Comment (xzandrew) [inline cmd/cmdexport/export.go]: We already have package pkg/fs that contains CopyFile and CopyDir, no need to add another dependency.

File: Gopkg.toml [hunk 4]
```diff
@@ -72,3 +72,7 @@
 [[constraint]]
   name = "github.com/aws/aws-sdk-go"
   version = "1.14.21"
+
+[[constraint]]
+  name = "github.com/otiai10/copy"
```
Comment (xzandrew) [inline Gopkg.toml]: e already have package pkg/fs that contains CopyFile and CopyDir, no need to add another dependency.

File: cmd/cmdexport/export.go [hunk 5]
```diff
@@ -87,11 +90,36 @@ func newExport(opts Opts) (*export, error) {
 		s.Logger.Info("Starting export. ReprocessHistorical is false, will use incremental checkpoints if available.")
 	}
 
-	err = os.MkdirAll(s.Locs.State, 0777)
-	if err != nil {
-		return nil, fmt.Errorf("could not create dir to save state, err: %v", err)
-	}
+	if fileutil.FileExists(s.Locs.State) {
```
Comment (xzandrew) [inline cmd/cmdexport/export.go]: I think for this it would be better to use fs.Exists that returns error as well, so we could make sure the value is correct if error is returned.

File: cmd/cmdexport/export.go [hunk 6]
```diff
@@ -87,11 +90,36 @@ func newExport(opts Opts) (*export, error) {
...
+
+		// if there is a tmp processed file and checkpoint dir, it means that the
+		// export was killed. If so, delete the current processed file and checkpoint
+		// dir and copy the those temps.
+		// if there are no temps, create them and defer delete
 
+		if fileutil.FileExists(s.Locs.LastProcessedFileTmp) {
```
Comment (xzandrew) [inline cmd/cmdexport/export.go]: Doing this here is ok for now, but there is a small downside that export could be cancelled after we remove Tmp dirs in defer but before the result is returned to service run.

Comment (xzandrew) [reply cmd/cmdexport/export.go]: This doesn't handle the case of cancelled first historical export, we would have unfinished state in that case and no tmp/backup.

File: pkg/iloader/integration.go [hunk 7]
```diff
@@ -114,10 +114,9 @@ func devIntegrationCommand(binaryName string) (*exec.Cmd, error) {
 
 	packageName := "github.com/pinpt/agent/integrations/" + binaryName
```
Comment (xzandrew) [inline pkg/iloader/integration.go]: Are these changes needed with your latest kill code? If not, revert back to run.

File: cmd/cmdexport/export.go (hunk 6 above)
Comment (xzandrew) [reply cmd/cmdexport/export.go]: It could be better to move the new code into new functions instead of making new even larger.

File: pkg/forcekill/forcekill.go (hunk 1 above)
Comment (pec1985) [reply pkg/forcekill/forcekill.go]: yes, you're right. I need to change it

File: cmd/cmdexport/export.go [hunk 8]
```diff
@@ -204,13 +174,72 @@ func newExport(opts Opts) (*export, error) {
 	return s, nil
 }
 
-func (s *export) Destroy() {
+func (s *export) createStateDir() error {
+	if stateExists, err := fs.Exists(s.Locs.State); stateExists {
+
```
Comment (xzandrew) [inline cmd/cmdexport/export.go]: The code would be more readable if you check err first and return if it's there, no need for nested brackets in that case.

File: cmd/cmdexport/export.go [hunk 9]
```diff
@@ -204,13 +174,72 @@ func newExport(opts Opts) (*export, error) {
...
+		// if there is a tmp processed file and checkpoint dir, it means that the
```
Comment (xzandrew) [inline cmd/cmdexport/export.go]: in comment tmp->backup

File: cmd/cmdexport/export.go [hunk 10]
```diff
@@ -204,13 +174,72 @@ func newExport(opts Opts) (*export, error) {
...
 	}
+	var err1, err2 error
+	if err := os.RemoveAll(s.Locs.LastProcessedFileBackup); err != nil {
+		err1 = fmt.Errorf("error deleting last processed backup file. err %v", err)
+	}
+	if err := os.RemoveAll(s.Locs.RipsrcCheckpointsBackup); err != nil {
+		err2 = fmt.Errorf("error deleting last checkpoints backup dir. err %v", err)
+	}
+	if err1 != nil || err2 != nil {
+		return fmt.Errorf("err1: %v. err2:%v", err1, err2)
```
Comment (xzandrew) [inline cmd/cmdexport/export.go]: you don't really need to return two errors. returning on the first would be ok and make the code simpler
